servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CapturedContextProviders.java

```diff
@@ -43,20 +41,32 @@ static List<CapturedContextProvider> providers() {
         return PROVIDERS;
     }
 
-    // TODO: this was copied from `ServiceLoaderUtils` because the `CapturedContextProvider` interface is package
-    //  private and the call to `ServiceLoader.load(..)` can only load implementations for types that are visible
-    //  to the package from which it is called (because it checks if it's accessible by checking stack trace).
-    //  One we make CapturedContextProvider public we can return to using ServiceLoaderUtils.
-    private static <T> List<T> loadProviders(final Class<T> clazz, final ClassLoader classLoader, final Logger logger) {
+    // This was copied from `ServiceLoaderUtils` because the implementation there logs the result, which normally
+    // wouldn't be a problem but as it turns out things like the MDC logging utils depend on the AsyncContext which
+    // may result in a cyclical dependency in class initialization.
+    private static <T> List<T> loadProviders(final Class<T> clazz, final ClassLoader classLoader) {
         final List<T> list = new ArrayList<>(0);
         for (T provider : ServiceLoader.load(clazz, classLoader)) {
             list.add(provider);
         }
         if (list.isEmpty()) {
-            logger.debug("ServiceLoader {}(s) registered: []", clazz.getSimpleName());
+            runLater(() -> getLogger().debug("ServiceLoader {}(s) registered: []", clazz.getSimpleName()));
             return emptyList();
         }
-        logger.info("ServiceLoader {}(s) registered: {}", clazz.getSimpleName(), list);
+        runLater(() -> getLogger().info("ServiceLoader {}(s) registered: {}", clazz.getSimpleName(), list));
         return unmodifiableList(list);
     }
+
+    // We have to run logging code 'later' because of cyclical dependency issues that can arise if logging depends
+    // on the AsyncContext.
+    private static void runLater(Runnable runnable) {
+        Thread t = new Thread(runnable);
+        t.setDaemon(true);
+        t.setName(CapturedContextProviders.class.getSimpleName() + "-logging");
+        t.start();
+    }
+
+    private static Logger getLogger() {
+        return LoggerFactory.getLogger(CapturedContextProviders.class);
```

Comment:
Would be great to try writing a reproducer, but if it's too involved then it's fine without it.

Reply:
Done in https://github.com/apple/servicetalk/pull/3202/commits/33dca1515039e903c0e8b726e5328137ef28265f. To reproduce we need to have both the MDC tools _and_ add the DynamicThresholdFilter.